### PR TITLE
Add segment tree solution for Load Balancing (usaco 624) 

### PR DESCRIPTION
**Pull request by ChuangSheep**
Add segment tree solution with cpp and java implementation. 
Refactor the binary search solution. 
Add explanations. 

_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making the changes.

### CODE REVIEW

**Comment by ChuangSheep** [comment]
ah btw closes #2793

File: solutions/usaco-624.mdx [hunk 1]
```diff
@@ -2,285 +2,378 @@
...
+The problem can be solved by sweeping a line from bottom to top. This line
+divides the farm into two regions, north and south. At the beginning, all cows
+belong to the northern region. For each of these lines, we need an efficient way
+to determine the optimal west/east partition to divide the farm into four pieces
+with the maximum number of cows appearing in one of the four regions, $M$, being
+minimized. To achieve this, we store the positions of cows in the northern and
+southern regions in two separate segment trees. These two segment trees hold the
+number of cows on a certain $x$ coordinate which belong to one of the two
+regions, respectively. Then, we do a binary search on both segment trees to find
+the best vertical line which minimizes $M$. In particular, we begin from the root
```
Comment (SansPapyrus683) [inline solutions/usaco-624.mdx:23]: what's "the root"?

File: solutions/usaco-624.mdx [hunk 2]
```diff
@@ -2,285 +2,378 @@
...
+and always add either the left or right subtree to the western or eastern
+region, depending on which of these two options leads to the smallest increase
+of $M$. At the end, we update our overall best solution and sweep the line on
+step to the top. We add all cows with current $y$ coordinate to the southern
+region and remove them from the northern region.
+
+A coordination compression of the $x$ coordinates is also performed at the
+beginning to constraint the size of our segment trees to be no larger than the
+number of cows.
```
Comment (SansPapyrus683) [inline solutions/usaco-624.mdx:32]: maybe give the more obvious reason that making a 1 billion size segtree is unfeasible lol

File: solutions/usaco-624.mdx [hunk 3]
```diff
@@ -2,285 +2,378 @@
...
+with the maximum number of cows appearing in one of the four regions, $M$, being
+minimized. To achieve this, we store the positions of cows in the northern and
+southern regions in two separate segment trees. These two segment trees hold the
+number of cows on a certain $x$ coordinate which belong to one of the two
+regions, respectively. Then, we do a binary search on both segment trees to find
+the best vertical line which minimizes $M$. In particular, we begin from the root
+and always add either the left or right subtree to the western or eastern
+region, depending on which of these two options leads to the smallest increase
+of $M$. At the end, we update our overall best solution and sweep the line on
+step to the top. We add all cows with current $y$ coordinate to the southern
```
Comment (SansPapyrus683) [inline solutions/usaco-624.mdx:27]: "sweep the line on the step to the top"
maybe "move the line to the next y-coordinate of interest"

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/usaco-624.mdx (hunk 1 above)
Comment (ChuangSheep) [reply solutions/usaco-624.mdx]: the root of the segment tree

File: solutions/usaco-624.mdx (hunk 2 above)
Comment (ChuangSheep) [reply solutions/usaco-624.mdx]: lol the constraint is 1 million which is, actually, feasible, without the coordinate compression

File: solutions/usaco-624.mdx [hunk 4]
```diff
@@ -2,285 +2,378 @@
...
+
+The problem can be solved by sweeping a line from bottom to top. This line
+divides the farm into two regions, north and south. At the beginning, all cows
+belong to the northern region. For each of these lines, we need an efficient way
+to determine the optimal west/east partition to divide the farm into four pieces
+with the maximum number of cows appearing in one of the four regions, $M$, being
+minimized. To achieve this, we store the positions of cows in the northern and
+southern regions in two separate segment trees. These two segment trees hold the
+number of cows on a certain $x$ coordinate which belong to one of the two
+regions, respectively. Then, we do a binary search on both segment trees to find
```
[the remark this answers is not in the document]
Comment (ChuangSheep) [reply solutions/usaco-624.mdx]: yea basically

File: solutions/usaco-624.mdx (hunk 2 above)
Comment (SansPapyrus683) [reply solutions/usaco-624.mdx]: would it still TLE?
if not then just say like "a segtree of size 1 million is kinda sus"

File: solutions/usaco-624.mdx [hunk 5]
```diff
@@ -2,285 +2,355 @@
...
+		 * partition between west and east by performing a binary search
+		 */
+		M = min(M, north.query(south));
+		// move the sweeping line to the next y coordinate of interest
+		i = j;
 	}
-	assert(ind == sz(v));
-	return bound;
+
+	cout << M << "\n";
```
Comment (SansPapyrus683) [inline solutions/usaco-624.mdx:167]:
```suggestion
	cout << M << endl;
```

File: solutions/usaco-624.mdx (hunk 2 above)
Comment (ChuangSheep) [reply solutions/usaco-624.mdx]: I have removed coordinate compression completely and the got ACC

File: solutions/usaco-624.mdx [hunk 6]
```diff
@@ -2,285 +2,355 @@
...
-}
-tcTU> T lstTrue(T lo, T hi, U f) {
-	lo --; assert(lo <= hi); // assuming f is decreasing
-	while (lo < hi) { // find first index such that f is true
-		T mid = lo+(hi-lo+1)/2;
-		f(mid) ? lo = mid : hi = mid-1;
+struct SegmentTree {
+	int size;
+	vector<int> tree;
+	SegmentTree() : size(1 << 20) { tree.assign(size * 2, 0); }
```
Comment (SansPapyrus683) [inline solutions/usaco-624.mdx:52]: i'd give a size as the argument instead of just making a really big one, up to you though

Comment (ChuangSheep) [reply solutions/usaco-624.mdx:52]: not possible unless we track the coordinates of the cows and get the maximum (remember we threw the coordinate compression away), and then find the next power of two. 
Or alternatively give the SegTree a size as parameter but then by initialization I m gonna have to put 1 << 20 in it, so doesn't really make sense.

File: solutions/usaco-624.mdx [hunk 7]
```diff
@@ -2,285 +2,355 @@
...
-	if (sz(s)) setIn(s+".in"), setOut(s+".out"); // for USACO
-}
+int main() {
+	freopen("balancing.in", "r", stdin);
+	freopen("balancing.out", "w", stdout);
+
+	int N;
+	cin >> N;
+	vector<pair<int, int>> cows(N);
+	for (auto &c : cows) {
```
Comment (SansPapyrus683) [inline solutions/usaco-624.mdx:124]:
```suggestion
	for (pair<int, int>&c : cows) {
```

File: solutions/usaco-624.mdx [hunk 8]
```diff
@@ -2,285 +2,355 @@
...
-	F0R(x,N) {
-		while (ind < sz(v) && v[ind].f == x) {
-			int Y = v[ind].s;
-			cnt[Y] ++; if (Y <= y) sum ++;
-			ind ++;
+	// sort after the y coordinates and divide the farm into north and south
+	sort(cows.begin(), cows.end(),
+		[&](const pair<int, int> &c1, const pair<int, int> &c2) {
+			return make_pair(c1.second, c1.first) < make_pair(c2.second, c2.first);
+		});
```
Comment (SansPapyrus683) [inline solutions/usaco-624.mdx:132]:
```suggestion
	sort(
		cows.begin(), cows.end(),
		[&](const pair<int, int> &c1, const pair<int, int> &c2) {
			return make_pair(c1.second, c1.first) < make_pair(c2.second, c2.first);
		}
	);
```

Comment (SansPapyrus683) [reply solutions/usaco-624.mdx:132]: i don't know

File: solutions/usaco-624.mdx [hunk 9]
```diff
@@ -357,3 +427,166 @@ public class Balancing {
...
 </LanguageSection>
+</Spoiler>
+
+
+## Solution 2: With Binary Search
+
+### Explanation
+
+Alternatively, we can also just binary search on the answer without using a PURQ
+data structure. To check if a given $M^*$ value is feasible, we calculate the
```
Comment (SansPapyrus683) [inline solutions/usaco-624.mdx:438]: what's M^*?

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/usaco-624.mdx [hunk 10]
```diff
@@ -437,17 +437,17 @@ public class Balancing {
...
-$M^*$ cows in the region. We then go through every $x$ coordinate and check if
-the sum of the heights of the southern and northern regions is larger than the
-height of the farm. As there are each two regions in the north and south, we
-always take the minimum between north-west/north-east and south-west/south-east.
-If there is such an $x$ partition that leads to the sum being larger than the
-height of the farm, then there is one solution with the current $x$ and $y$
-partition.
-
-We do a coordinate compression on the $x$ and $y$ coordinates to simplify the 
+data structure. To check if a given $M^*$, the maximum number of cows in any of
```
[the remark this answers is not in the document]
Comment (ChuangSheep) [reply solutions/usaco-624.mdx:440]: `M^*` is only a variable, introduced and defined as the maximum number of cows in any of the four regions. As this variable should not be confused with the real `M` defined in the problem statement, it is common to add an asterisk on top to make the distinction. We don't raise anything.

[the remark this answers is not in the document]
Comment (SansPapyrus683) [reply solutions/usaco-624.mdx:440]: then i'd make it clear that this is a completely different variable- "we'll check if [asdfasdf], henceforth referred to as $M^*$..."

**Review by SansPapyrus683 — APPROVED** [review]
yeah

File: solutions/usaco-624.mdx [hunk 11]
```diff
@@ -357,3 +429,168 @@ public class Balancing {
...
+		if (cows[i].second != actual_y) {
+			new_y++;
+			actual_y = cows[i].second;
+		}
+		cows[i].second = new_y;
+	}
+	//EndCodeSnip
+
+	sort(cows.begin(), cows.end());
+	int l = 1, h = N + 1;
```
Comment (ryanchou-dev) [inline solutions/usaco-624.mdx:545]: perhaps rename to `l, r`?

doesn't rly matter tho

**Review by ryanchou-dev — APPROVED** [review]
wooo

**Pull request merged**